src/transport/PeerMessageCounter.h

```diff
@@ -0,0 +1,189 @@
+/*
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines the CHIP message counters of remote nodes.
+ *
+ */
+#pragma once
+
+#include <array>
+#include <bitset>
+
+#include <support/Span.h>
+
+namespace chip {
+namespace Transport {
+
+class PeerMessageCounter
+{
+public:
+    static constexpr size_t kChallengeSize = 8;
+
+    PeerMessageCounter() : mStatus(Status::NotSynced) {}
+    ~PeerMessageCounter() { Reset(); }
+
+    void Reset()
+    {
+        switch (mStatus)
+        {
+        case Status::NotSynced:
+            break;
+        case Status::SyncInProcess:
+            mSyncInProcess.~SyncInProcess();
+            break;
+        case Status::Synced:
+            mSynced.~Synced();
+            break;
+        }
+        mStatus = Status::NotSynced;
+    }
+
+    bool IsSynchronizing() { return mStatus == Status::SyncInProcess; }
+    bool IsSynchronized() { return mStatus == Status::Synced; }
+
+    void SyncStarting(FixedByteSpan<kChallengeSize> challenge)
+    {
+        assert(mStatus == Status::NotSynced);
+        mStatus = Status::SyncInProcess;
+        new (&mSyncInProcess) SyncInProcess();
+        mSyncInProcess.mChallenge = challenge.ToArray();
```

Comment:
So what this does in practice is allocate a new `std::array` on the stack, do `ToArray` into it, then do the copy, as far as I can tell.

If we had mSyncInProcess.mChallenge be `uint8_t[kChallengeSize]` and just used `memcpy` here, we would get smaller code, afaict.  See 
* https://godbolt.org/z/5ohfKnzTs ARM 64-bit Clang 9
* https://godbolt.org/z/Yxrfnd74x ARM 32-bit Clang 11
* https://godbolt.org/z/cajjPMEzf ARC 32-bit GCC 10 

all with -Os.  It looks to me like using a C-type array instead of `std::array` is a win in all those cases as far as codesize goes.